comptime.md

```diff
@@ -0,0 +1,123 @@
+    $if T.kind == .int {
+        return int_string(val)
+    } else if T.kind == .array {
+        mut out := '['
+        for i,elem in val {
+            out = '${out}${elem}'
+            if i != val.len - 1 {
+                out = '${out},'
+            }
+        }
+        return '${out}]'
+    } else if T.kind == .struct_ {
+        out = '${$T.name} {'
+        $for field in T.fields {
+            field := val.$field 
+            out = '${out} ${$field}: ${field}'
+        }
+        return '${out}}'
+    } else {
+        // and so on...
+    }
+}
+```
+
+Notice that we do not have `$else` anymore, that is because with this proposal if parser sees `$` it simply continues parsing V statements or expressions *but* parsed statement and expression will be marked with `comptime` flag.
+
+And here is simple example of factorial that is executed fully at compile-time! 
+```v
+fn factorial(x int) int {
+    if x < 2 {
+        return x 
+    } 
+
+    return factorial(x - 1) * x
+}
+
+fn main() {
+    cx := $factorial(5) // call prefixed with `$` becomes compile-time call
+    $cy := factorial(6) // variable assignment prefixed with `$` puts it to compile-time environment 
+    println(cx) // 120 
+    println($cy) // 720
+}
+```
+
+# Reference-level explanation 
+
+In parsing we should change our handling of `$` symbol. Instead of parsing specific comptime cases we should simply continue parsing expression or statement but in the end mark it with `comptime`. `comtpime` mark can be stored in each of AST nodes. 
+
+To implement this we would not need to significantly change checker, we would just need to add additional compile-time environment for comptime variables (generics will be treated as comptime variables too!). After code is checked we would just continue to codegen and each of `v.gen.*` implementations *must* evaluate expressions and statements marked with `comptime`. This is where it might get complex but to solve this we will simply use `v.eval`, `v.eval.Eval` will be stored per generation context. We should also find a way to pass generation context to eval context so we can for example implement `include` builtin call that would include C/JS file into output file. Something like this:
+
+```v
+gen.eval.add_func('include', fn (gen cgen.Gen, args []Value) {
+    gen.includes.write_string(args[0].str())
+})
+
+// used like this in toplevel:
+$include("header.h")
+```
+
+# Drawbacks 
+This would break a lot of code that depends on current implementation of compile-time and it would take quite a lot of time to switch to new comptime and implement it. 
+
+# Rationale and alternatives 
+
+This design allows to execute entire V language at comptime and it makes it possible to:
+- Implement serializers/deserializers
+- Write code that would produce new code a-la Lisp macros 
+- Write code that would generate new types for example generating Matrix struct with demensions known at compile-time.
+
```

Comment:
I'd add two more important points (the first being IMHO **enormously** important):

1. Overall safety. There has be numerous discussions about compile-time assert, ahead-of-time proving of correctness of certain properties of passed data to a function etc. Different compile-time warnings, etc.

2. Optimization. Proving certain aspects of given data etc. helps a lot with compile-time optimization. Imagine the regex module we have compiling vast majority of expressions in compile-time to highly efficient bytecode. Now this has to be done in run time without any good way to check for correctness of the regex syntax itself as well as without any opportunity of optimization of register spilling, caching etc.